### PR TITLE
feat: Added Contextual Tuples to Assertions API

### PR DESCRIPTION
**Pull request by smeadows-okta**
Added support for contextual tuple to assertions

## Description
Currently, assertions are pure (object, relation, subject) tuples. Added support for contextual tuples to assertions for advanced scenarios.

## References
[issue-946](https://github.com/openfga/openfga/issues/946)

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/server/test/write_assertions.go [hunk 1]
```diff
@@ -208,3 +250,134 @@ type repo
 		})
 	}
 }
+
+func TestConceptualTuplesWriteAssertionsFailure(t *testing.T, datastore storage.OpenFGADatastore) {
```
Comment (miparnisari) [inline pkg/server/test/write_assertions.go:254]: In the previous test you added elements to the test case array. 👍 

But here you basically copied `TestWriteAssertionsFailure`. Why? You could have done what you did above, and include your new test cases in `TestWriteAssertionsFailure`

File: pkg/server/test/write_assertions.go [hunk 2]
```diff
@@ -208,3 +250,134 @@ type repo
...
+	}
+	ctx := context.Background()
+
+	writeAuthzModelCmd := commands.NewWriteAuthorizationModelCommand(datastore)
+	modelID, err := writeAuthzModelCmd.Execute(ctx, githubModelReq)
+	require.NoError(t, err)
+
+	var tests = []writeAssertionsTestSettings{
+		{
+			_name: "write_conceptual_tuple_assertion_with_invalid_relation_fails",
```
Comment (miparnisari) [inline pkg/server/test/write_assertions.go:284]:
```suggestion
			_name: "write_contextual_tuple_assertion_with_invalid_relation_fails",
```

here and elsewhere

File: pkg/server/test/write_assertions.go [hunk 3]
```diff
@@ -208,3 +250,134 @@ type repo
...
+					Expectation: false,
+				},
+			},
+			modelID: modelID.GetAuthorizationModelId(),
+			err: serverErrors.ValidationError(
+				fmt.Errorf("invalid 'object' field format"),
+			),
+		},
+		{
+			_name: "write_conceptual_tuple_assertion_with_invalid_object_fails",
```
Comment (miparnisari) [inline pkg/server/test/write_assertions.go:326]:
```suggestion
			_name: "write_conceptual_tuple_assertion_with_invalid_user_fails",
```

File: pkg/server/test/write_assertions.go [hunk 4]
```diff
@@ -208,3 +250,134 @@ type repo
...
+						},
+					},
+					Expectation: false,
+				},
+			},
+			modelID: "not_valid_id",
+			err: serverErrors.AuthorizationModelNotFound(
+				"not_valid_id",
+			),
+		},
```
Comment (miparnisari) [inline pkg/server/test/write_assertions.go:367]: This test isn't useful - validation of contextual happens will never happen if the model ID isn't valid.

**Review by miparnisari — CHANGES_REQUESTED** [review]

File: pkg/server/test/write_assertions.go (hunk 4 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go]: 👍

File: pkg/server/test/write_assertions.go (hunk 3 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go:331]: 👍

File: pkg/server/test/write_assertions.go (hunk 2 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go:287]: 👍

File: pkg/server/test/write_assertions.go (hunk 1 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go]: I broke them out into separate test cases tuples vs conceptual tuples. Each are handled differently in the code (Check, WriteAssertion) and returns the same error codes and need to be tested as such and allow anyone to expand on them without impact the other.

Comment (miparnisari) [reply pkg/server/test/write_assertions.go]: > tuples vs contextual tuples. 

Both are part of the WriteAssertion input. 

In general, across the codebase, we have test cases with a `request` input that tests all the various possible combinations for any API. I think this specific suite shouldn't be any different.

> Each are handled differently in the code (Check, WriteAssertion) and returns the same error codes

If we want, in the future we can come back and create a helper utility that has all the contextual tuple validation code.

**Review by miparnisari — APPROVED** [review]

File: pkg/server/test/write_assertions.go (hunk 1 above)
Comment (smeadows-okta) [reply pkg/server/test/write_assertions.go]: 👍

**Pull request merged**